openssl-sys/src/openssl_shim.c

```diff
@@ -3,6 +3,33 @@
 #include <openssl/dh.h>
 #include <openssl/bn.h>
 
+#if defined(__APPLE__) || defined(__linux)
+
+#include<pthread.h>
+#include<openssl/crypto.h>
+
+unsigned long thread_id()
+{
+    unsigned long ret = (unsigned long)pthread_self();
+    return ret;
+}
+
+void rust_openssl_set_id_callback() {
+  CRYPTO_set_id_callback((unsigned long (*)())thread_id);
+}
+
+#else
+// Openssl already handles Windows directly, so we don't
+// need to explicitly set it
+
+void rust_openssl_set_id_callback() {
+  // We don't know how to set the callback for arbitrary OSes
```

Comment:
We could. However, `pthread_self()` returns a `pthread_t`, which is opaque (according to the posix spec). On Linux and OSX it's okay to cast it to ulong the way we do, but for a random POSIX OS there's no guarantee.

I don't mind converting this to an `#ifdef __posix`, though. It's *probably* correct for common posix OSes.